Add support for stateless join on NULL key

kevincherian opened the pull request:
Summary
---------

Issue: https://github.com/fentik/flink/issues/30

This pull request adds optimization for outer joins on a NULL key - while emitting the complete state, it evicts null records for these queries. As mentioned in https://github.com/fentik/flink/issues/30, Flink Join operators maintain
a complete set of all input data from left & right sides during a join. In the case of an outer equi join on a null key, we don't need to maintain as much state as the NULL value in the key will result in a failing join condition - in these cases, we can safely construct output rows without maintain any op state. This pull request builds this optimization for "Batch Backfill" mode - where when switching from batch to streaming (i.e. emitStateAndSwitchToStreaming), it cleans up state
for null keys -- thereby reducing operator state that is maintained. This PR also adds an execution option called
table.exec.stateless-join-on-null-key that allows you to control whether this optimization is enabled.

Test Plan
-------------
1) Execute following commands in Flink SQL: set table.exec.stateless-join-on-null-key='True'; set table.exec.batch-backfill='True';
2) Create NULL JOIN key in Postgress: update product set category_id=NULL where id=5699;
3) Execute LEFT OUTER JOIN in Flink SQL: select pc.id as category_id, p.id as product_id, product_name, category_name from product p left outer join product_category pc on p.category_id=pc.id;
4) Tail logs and make sure log line:
   "Performance Optimization - Evicting NULL record from state in outer
equijoin" is present
5) Run internal fentik regression test pipeline

See <img width="1767" alt="Screen Shot 2023-03-21 at 3 25 55 PM" src="https://user-images.githubusercontent.com/8899705/226766232-6a189ea6-b107-4f74-8f5c-05cac4576cf0.png">
where we see records get evicted from state

Commit:
Add support for stateless left join on NULL key

Summary
---------

Issue: https://github.com/fentik/flink/issues/30

This pull request adds optimization for outer joins on a NULL key - it
evicts records while emitting complete by evicting null records for
these queries. As mentioned in
https://github.com/fentik/flink/issues/30, Flink Join operators maintain
a complete set of all input data from left & right sides during a join.
In the case of an outer equi join on a null key, we don't need to
maintain as much state as the NULL value in the key will result in a
failing join condition - in these cases, we can safely construct output
rows without maintain any op state. This pull request builds this
optimization for "Batch Backfill" mode - where when switching from batch
to streaming (i.e. emitStateAndSwitchToStreaming), it cleans up state
for null keys -- thereby reducing operator state that is maintained.

This PR also adds an execution option called
table.exec.stateless-join-on-null-key that allows you to control whether
this optimization is enabled.

Test Plan
-------------
1) Execute following commands in Flink SQL: a) set
   table.exec.stateless-join-on-null-key='True'; set
table.exec.batch-backfill='True'
1) Create NULL JOIN key in Postgress: update product set category_id=NULL where id=5699;
2) Execute LEFT OUTER JOIN in Flink SQL: select pc.id as category_id, p.id as product_id, product_name, category_name from product p left outer join product_category pc on p.category_id=pc.id;
3) Tail logs and make sure log line:
   "Performance Optimization - Evicting NULL record from state in outer
equijoin" is present
4) Run internal fentik regression test pipeline

sergeitsar commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -369,6 +369,14 @@ public class ExecutionConfigOptions {
+	public static final ConfigOption<Boolean> TABLE_EXEC_STATELESS_JOIN_ON_NULL_KEY =
+			key("table.exec.stateless-join-on-null-key")
+                    .booleanType()
+                    .defaultValue(false)
```
let's flip this to default on for our pipelines

sergeitsar commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperator.java:
```diff
@@ -186,17 +189,22 @@ public void processElement1(StreamRecord<RowData> element) throws Exception {
                 // other side to pass through
-                // RowDataStringSerializer rowStringSerializer = new RowDataStringSerializer(leftType);
-                // LOG.info("SERGEI left input optimization {}", rowStringSerializer.asString(element.getValue()));
+                RowDataStringSerializer rowStringSerializer = new RowDataStringSerializer(leftType);
```
as part of this change, let's move the row string serializer creation into open() -- make one for left and one for right rows, and use them consistently -- creating these for every printed row is a vestigial leftover from a prior refactor I didn't finish in join

sergeitsar commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperator.java:
```diff
@@ -186,17 +189,22 @@ public void processElement1(StreamRecord<RowData> element) throws Exception {
                 leftInputDroppedNullKeyCount.inc();
                 return;
-            } else if (isEquijoin && !isBatchMode() && statelessNullKeysEnabled) {
+        } else if (StreamingPerfOptimizationUtil.isEligibleForNullEquijoinOptimization(
```
indentation looks off

sergeitsar commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperator.java:
```diff
@@ -221,15 +229,22 @@ public void processElement2(StreamRecord<RowData> element) throws Exception {
                 // we will never have associated rows data from the
                 // other side to pass through
+                RowDataStringSerializer rowStringSerializer = new RowDataStringSerializer(rightType);
+                LOG.debug("Performance Optimization - Right outer join on null key {}", rowStringSerializer.asString(element.getValue()));
```
the print statement is misleading: the input here is the whole row and not the key

sergeitsar commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperator.java:
```diff
@@ -186,17 +189,22 @@ public void processElement1(StreamRecord<RowData> element) throws Exception {
-                // RowDataStringSerializer rowStringSerializer = new RowDataStringSerializer(leftType);
-                // LOG.info("SERGEI left input optimization {}", rowStringSerializer.asString(element.getValue()));
+                RowDataStringSerializer rowStringSerializer = new RowDataStringSerializer(leftType);
+                LOG.debug("Performance Optimization - Left outer join on null key {}", rowStringSerializer.asString(element.getValue()));
```
the print statement is misleading: the input here is the whole row and not the key

sergeitsar commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperator.java:
```diff
@@ -289,27 +304,26 @@ protected boolean isHybridStreamBatchCapable() {
         if (leftIsOuter) {
             leftRecordStateView.emitCompleteState(getKeyedStateBackend(), this.collector,
-                    rightRecordStateView, joinCondition, false, true /* inputIsLeft */);
+                    rightRecordStateView, joinCondition, false, true, isEquijoin, true, statelessNullKeysEnabled, isBatchMode());
```
you have isBatchMode plumbed through the APIs, but I didn't see actually get used (when we're emitting, we know that we're going from batch -> streaming, so when this is called, batchmode is always true

kevincherian replied:
Good point - addressed in latest commit

sergeitsar commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/OuterJoinRecordStateViews.java:
```diff
@@ -160,6 +161,17 @@ static void processElements(JoinRecordStateView otherView, Collector<RowData> co
         }
+
+        if (!isAntiJoin && StreamingPerfOptimizationUtil.isEligibleForNullEquijoinOptimization(
+                StreamingPerfOptimizationUtil.isKeyAnyNulls(thisRow),
```
thisRow is the complete row here, not just the key -- you'll be incorrectly capturing non-key payload columns which are null but the key itself is not null, which is incorrect
the right place to check for "are there nulls in the key for this row" is in the loop that iterates keys in the backend right next to the call that sets backend key context, you alternatively you pass the join key into there

kevincherian replied:
Ah good catch! Thank you - fixed!

sergeitsar commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/OuterJoinRecordStateViews.java:
```diff
@@ -160,6 +161,17 @@ static void processElements(JoinRecordStateView otherView, Collector<RowData> co
+                true)) {
+            // Retract record from state if it is an outer equijoin on a null key (Reduces operator state we need to maintain)
+            LOG.debug("Performance Optimization - Evicting NULL record from state in outer equijoin");
+            thisView.retractRecord(thisRow);
```
let's add counters here: both for number of records emitted as well as number of records removed from state

kevincherian replied:
done

Commit:
Address CR comments

1) Fix bad indent
2) Add counters for different workflows - numRowsEmitted,
   numRowsEvictedDuringEmit,
leftInputNullOuterEquiJoinOptimizationCount, rightInputNullOuterEquiJoinOptimizationCount
3) Fix logs with incorrect messages
4) Remove unused isBatchMode
5) When checking for null equijoin optimization, ensuring that you are
   using only the key & not the full "record" to avoid bug of any null
col triggering the optimization

Commit:
Run spotless + evict state even during anti-join switch to streaming

Commit:
Fix bug where deletes were not being set with correct rowKind during null optimization

sergeitsar reviewed (approved):
lgtm, ship it!
